src/main/java/org/junitpioneer/jupiter/resource/ResourceExtension.java

```diff
@@ -0,0 +1,368 @@
+		store.put(uniqueKey(), resourceFactory);
+		Resource<?> resource;
+		try {
+			resource = resourceFactory.create(unmodifiableList(asList(newAnnotation.arguments())));
+			store.put(uniqueKey(), resource);
+		}
+		catch (Exception ex) {
+			throw new ParameterResolutionException(
+				"Unable to create a resource from `" + resourceFactory.getClass() + '`', ex);
+		}
+		try {
+			return resource.get();
+		}
+		catch (Exception ex) {
+			throw new ParameterResolutionException(
+				"Unable to get the contents of the resource created by `" + resourceFactory.getClass() + '`', ex);
+		}
+	}
+
+	private Object resolveShared(Shared sharedAnnotation, Parameter[] parameters, ExtensionContext.Store store) {
+		throwIfHasAnnotationWithSameNameButDifferentType(store, sharedAnnotation);
+		throwIfMultipleParametersHaveExactAnnotation(parameters, sharedAnnotation);
+
+		ResourceFactory<?> resourceFactory = store
+				.getOrComputeIfAbsent( //
+					factoryKey(sharedAnnotation), //
+					__ -> ReflectionSupport.newInstance(sharedAnnotation.factory()), //
+					ResourceFactory.class);
+		ResourceWithLock<?> resource;
+		try {
+			resource = store
+					.getOrComputeIfAbsent( //
+						resourceKey(sharedAnnotation), //
+						__ -> createResource(sharedAnnotation, resourceFactory), //
+						ResourceWithLock.class);
+		}
+		catch (UncheckedParameterResolutionException e) {
+			throw e.getCause();
+		}
+		try {
+			return resource.get();
+		}
+		catch (Exception e) {
+			throw new ParameterResolutionException(
+				"Unable to get the contents of the resource created by `" + sharedAnnotation.factory() + '`', e);
+		}
+	}
+
+	private ResourceWithLock<?> createResource(Shared sharedAnnotation, ResourceFactory<?> resourceFactory) {
+		try {
+			return new ResourceWithLock<>(
+				resourceFactory.create(unmodifiableList(asList(sharedAnnotation.arguments()))));
+		}
+		catch (Exception e) {
+			throw new UncheckedParameterResolutionException(new ParameterResolutionException(
+				"Unable to create a resource from `" + sharedAnnotation.factory() + "`", e));
+		}
+	}
+
+	private void throwIfHasAnnotationWithSameNameButDifferentType(ExtensionContext.Store store,
+			Shared sharedAnnotation) {
+		ResourceFactory<?> presentResourceFactory = //
+			store.getOrDefault(factoryKey(sharedAnnotation), ResourceFactory.class, null);
+
+		if (presentResourceFactory == null) {
+			store.put(keyOfFactoryKey(sharedAnnotation), factoryKey(sharedAnnotation));
+		} else {
+			String presentResourceFactoryName = //
+				store.getOrDefault(keyOfFactoryKey(sharedAnnotation), String.class, null);
+
+			if (factoryKey(sharedAnnotation).equals(presentResourceFactoryName)
+					&& !sharedAnnotation.factory().equals(presentResourceFactory.getClass())) {
+				throw new ParameterResolutionException(
+					"Two or more parameters are annotated with @Shared annotations with the name \""
+							+ sharedAnnotation.name() + "\" but with different factory classes");
```

Comment:
I see now that one of the tests is failing, I think because it's expecting this exception to be thrown but it's not.

I'll investigate when I next have the time.

Reply:
Failing test fixed.